### PR TITLE
Add pingback support (bug 1377305)

### PR DESCRIPTION
**Pull request by zalun**
Added POST API endpoint `/landings/{landing_id}/update`.
`landing_id` (from path) and `request_id` (from data) have to correspond
with database entry, otherwise 404 is raised.
Modified the schema by adding error and result.

### CODE REVIEW

File: landoapi/models/landing.py [hunk 1]
```diff
@@ -42,35 +51,45 @@ class Landing(db.Model):
     request_id = db.Column(db.Integer, unique=True)
     revision_id = db.Column(db.String(30))
     status = db.Column(db.Integer)
+    auth = db.Column(db.String(30))
```
Comment (mars-f) [inline landoapi/models/landing.py]: I don't think the auth mechanism should be stored in the database.  The [security design doc](https://docs.google.com/document/d/1cRIdpVqYcn0Mle7vyjJePNw5TcFoZ20N42dLVy5xVEk/edit#) says we should consider a shared API key for validating the client doing the update, and that means we'll end up adding a second migration at a later to remove the auth column.  I recommend we split this user story into two parts: one for this patch to perform status updates with no auth, and a second story to have a special API key for a pingback client to use to update statuses.

File: docker-compose.yml [hunk 2]
```diff
@@ -17,6 +17,7 @@ services:
       - PHABRICATOR_UNPRIVILEGED_API_KEY=api-123456789
       - TRANSPLANT_URL=https://stub.transplant.example.com
       - DATABASE_URL=sqlite:////db/sqlite.db
+      - HOST_URL=https://landoapi.mozilla.org
```
Comment (purelogiq) [inline docker-compose.yml]: Agree that we will need this variable. Lets set to `http://lando-api.test` and in ckolos change it for dev, stage, and prod.

File: landoapi/spec/swagger.yml [hunk 3]
```diff
@@ -121,6 +121,45 @@ paths:
...
+      operationId: landoapi.api.landings.update
+      description: |
+        Sends request to transplant service and responds just the status code
+      parameters:
+        - name: data
+          in: body
+          description: |
+            Retrieve status of the landing job
+          required: true
+          schema:
```
Comment (purelogiq) [inline landoapi/spec/swagger.yml]: See https://docs.google.com/document/d/1cRIdpVqYcn0Mle7vyjJePNw5TcFoZ20N42dLVy5xVEk/edit# for the remaining data.

File: landoapi/api/landings.py [hunk 4]
```diff
@@ -71,3 +71,18 @@ def get(landing_id):
...
+    if not landing:
+        return problem(
+            404,
+            'Landing not found',
+            'The requested Landing does not exist',
+            type='https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404'
+        )
+    landing.status = data['status']
+    landing.save()
+    return landing.serialize(), 202
```
Comment (purelogiq) [inline landoapi/api/landings.py]: Does transplant expect a 202, or 200? 
https://httpstatuses.com/202 vs https://httpstatuses.com/200

File: landoapi/models/landing.py [hunk 5]
```diff
@@ -1,10 +1,13 @@
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import os
+
 from landoapi.models.storage import db
 from landoapi.phabricator_client import PhabricatorClient
 from landoapi.transplant_client import TransplantClient
 
+TRANSPLANT_JOB_STARTING = 'starting'
```
Comment (purelogiq) [inline landoapi/models/landing.py]: Just for me, what is the difference between 'starting' and 'started'?

Comment (zalun) [inline landoapi/models/landing.py]: It is set before request is made. No idea if it is started on the Transplant side.

File: landoapi/api/landings.py (hunk 4 above)
Comment (zalun) [inline landoapi/api/landings.py]: I think it just checks for errors

File: landoapi/models/landing.py (hunk 5 above)
Comment (mars-f) [inline landoapi/models/landing.py]: > It is set before request is made. No idea if it is started on the Transplant side.

'pending' might be a better word to communicate the status.

File: landoapi/api/landings.py [hunk 6]
```diff
@@ -71,3 +69,25 @@ def get(landing_id):
...
+        return problem(
+            404,
+            'Landing not found',
+            'The requested Landing does not exist',
+            type='https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404'
+        )
+
+    landing.error = data.get('error_msg', '')
+    landing.result = data.get('result', '')
+    landing.status = TRANSPLANT_JOB_LANDED if data['landed'
```
Comment (mars-f) [inline landoapi/api/landings.py]: This line confuses me.  If the 'landed' field is present, regardless of value, everything is OK, and if an update is ever made with this field missing, it's a failure in the other system?  I think there should be a comment or two in here to explain what the possible combinations of incoming values are, and to explain what they represent in the other system and in our system.

File: landoapi/models/landing.py [hunk 7]
```diff
@@ -40,39 +49,34 @@ def create(cls, revision_id, phabricator_api_key=None, save=True):
...
         repo = phab.get_revision_repo(revision)
 
+        # save landing to make sure we've got the callback
+        landing = cls(
+            revision_id=revision_id,
+        ).save()
+
         trans = TransplantClient()
+        callback = '%s/landings/%s/update' % (
+            os.getenv('HOST_URL'), landing.id
```
Comment (mars-f) [inline landoapi/models/landing.py]: This setting's name could be improved.  I can't tell from the name alone that it is used for pingbacks.  Maybe something like `PINGBACK_HOST_URL` or just `PINGBACK_URL`?

File: landoapi/spec/swagger.yml [hunk 8]
```diff
@@ -111,8 +111,70 @@ paths:
...
+            properties:
+              id:
+                type: integer
+                description: |
+                  A newly created Landing id
         404:
-          description: Revision does not exist
+          description:
+            Revision does not exist or its request_id doesn't correspond to
+            the saved one
```
Comment (mars-f) [inline landoapi/spec/swagger.yml]: Probably best not to mention internal implementation details here.  I don't see a `request_id` mentioned in the POST parameters, either?

File: landoapi/spec/swagger.yml [hunk 9]
```diff
@@ -111,8 +111,70 @@ paths:
...
+        default:
+          description: Unexpected error
+          schema:
+            allOf:
+              - $ref: '#/definitions/Error'
+  /landings/{landing_id}/update:
+    post:
+      operationId: landoapi.api.landings.update
+      description: |
+        Sends request to transplant service and responds just the status code
```
Comment (mars-f) [inline landoapi/spec/swagger.yml]: Typo, "responds with just the status code"

File: landoapi/models/landing.py (hunk 7 above)
Comment (zalun) [inline landoapi/models/landing.py]: it's the lando-api host url - I thought we might use it for something else.

File: landoapi/spec/swagger.yml (hunk 8 above)
Comment (zalun) [inline landoapi/spec/swagger.yml]: right - it should stay as it was before
`request_id` is mentioned as "in body" parameter

File: landoapi/api/landings.py (hunk 6 above)
Comment (zalun) [inline landoapi/api/landings.py]: I'm adding description of fields which are received from Transplant (from Google doc)

File: landoapi/models/landing.py (hunk 7 above)
Comment (mars-f) [inline landoapi/models/landing.py]: Discussed on IRC, I'm fine with both keeping the variable as-is and with changing it.

Comment (purelogiq) [inline landoapi/models/landing.py]: I think HOST_URL works well, imo.

**Pull request merged**